src/main/steps/c100-rebuild/miam/navigationController.ts

```diff
@@ -0,0 +1,103 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import { Case } from '../../../app/case/case';
+import { MiamNonAttendReason } from '../../../app/case/definition';
+import {
+  C100_HEARING_URGENCY_URGENT,
+  C100_MIAM_CHILD_PROTECTION,
+  C100_MIAM_GENERAL_REASONS,
+  C100_MIAM_GET_MEDIATOR,
+  C100_MIAM_MIAM_DOMESTIC_ABUSE,
+  C100_MIAM_NO_NEED_WITH_REASONS,
+  C100_MIAM_OTHER,
+  C100_MIAM_OTHER_PROCEEDINGS,
+  C100_MIAM_PREVIOUS_ATTENDANCE,
+  C100_MIAM_URGENCY,
+  PageLink,
+} from '../../urls';
+
+class MIAMNavigationController {
+  protected selectedPages: MiamNonAttendReason[] = [];
+
+  private pages: Record<MiamNonAttendReason, Record<string, any>> = {
+    [MiamNonAttendReason.DOMESTIC]: {
+      url: C100_MIAM_MIAM_DOMESTIC_ABUSE,
+      dataReference: 'miam_domesticAbuse',
+    },
+    [MiamNonAttendReason.CHILD_PROTECTION]: {
+      url: C100_MIAM_CHILD_PROTECTION,
+      dataReference: 'miam_childProtectionEvidence',
+    },
+    [MiamNonAttendReason.URGENT]: {
+      url: C100_MIAM_URGENCY,
+      dataReference: 'miam_urgency',
+    },
+    [MiamNonAttendReason.PREV_MIAM]: {
+      url: C100_MIAM_PREVIOUS_ATTENDANCE,
+      dataReference: 'miam_previousAttendance',
+    },
+    [MiamNonAttendReason.EXEMPT]: {
+      url: C100_MIAM_OTHER,
+      dataReference: 'miam_notAttendingReasons',
+    },
+    [MiamNonAttendReason.NONE]: {
+      url: C100_MIAM_GET_MEDIATOR,
+    },
+  };
+
+  private getNextPageUrl(currentPageUrl: PageLink): PageLink | null {
+    let pageUrl: PageLink | null = null;
+    const currentPageId = Object.keys(this.pages).find(
+      pageId => this.pages[pageId].url === currentPageUrl
+    ) as MiamNonAttendReason;
+
+    if (currentPageId) {
+      const currentPageIndex: number = this.selectedPages.indexOf(currentPageId);
+
+      if (currentPageIndex < this.selectedPages.length - 1) {
+        pageUrl = this.pages[this.selectedPages[currentPageIndex + 1]].url;
+      }
+    }
+    return pageUrl;
+  }
+
+  public checkForAnyValidReason(caseData: Partial<Case>, page?: MiamNonAttendReason): boolean {
+    if (page && !this.selectedPages.includes(page)) {
+      return false;
+    }
+
+    const pages = page ? [page] : this.selectedPages;
+
+    return pages.some(pageId => {
```

Comment:
can we avoid this loop by setting the some flag on selection of reason

Reply:
This looping is necessary and time complexity is O(n) . The data type is an array and I have used a some constructor to check and stop the execution of the loop when the condition satisfies so it does not run the entire length of the loop. Also setting a flag on selection of reason cannot be done when we come back later using the dashboard.